### PR TITLE
Initial Classes

### PR DESCRIPTION
**Pull request by mecha**

### CODE REVIEW

File: src/AbstractBooking.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+abstract class AbstractBooking
+{
+    /**
+     * Retrieves the booking period.
+     *
+     * @since [*next-version*]
+     *
+     * @return PeriodInterface The period instance.
+     */
+    abstract protected function _getPeriod();
```
Comment (XedinUnknown) [inline src/AbstractBooking.php]: Given that `BookingInterface extends PeriodInterface`, is it the right choice to have an abstract booking *has* a period, when a concrete booking *is* a period?

File: src/StatusAwareTrait.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+     * @var string|Stringable
+     */
+    protected $status;
+
+    /**
+     * Retrieves the status associated with this instance.
+     *
+     * @since [*next-version*]
+     *
+     * @return string|Stringable The status.
```
Comment (XedinUnknown) [inline src/StatusAwareTrait.php]: Should `null` maybe be permissible?

File: src/AbstractBooking.php (hunk 1 above)
Comment (mecha) [reply src/AbstractBooking.php]: IMO, `BookingInterface` extending `PeriodInterface` signifies that a booking has the API of a period, and that it can be used as one. Does not necessarily mean that a booking _has_ to be a period under the hood.

This partial implementation does not assume that the booking itself is the period. It also does not imply that the booking will delegate to another period. This abstract method liberates extending classes by allowing them to do either.

In other words, since there is code in the abstract class that needs a `PeriodInterface` and since the abstract class cannot be sure whether its child classes will implement `PeriodInterface`, this point of customization was created.

File: src/StatusAwareTrait.php (hunk 2 above)
Comment (mecha) [reply src/StatusAwareTrait.php]: Ideally, yes. I'll make the change.

File: src/AbstractBooking.php (hunk 1 above)
Comment (XedinUnknown) [reply src/AbstractBooking.php]: No problem, if you have the functionality in traits, the different concrete implementation can `use` that.

I agree that this is flexible. However, conceptually every booking *is* a period. Therefore, I can much more easily imagine something that creates booking *from* periods, rather than a period being injected into a booking. So, maybe this flexibility is kinda for nothing. But good point about the abstract class not being able to guarantee that it is a period. Therefore, perhaps you may want to think about splitting the functionality into traits more.

Comment (mecha) [reply src/AbstractBooking.php]: No question about it. Conceptually, every booking _is_ a period. How they are constructed however does not relate to this. Created from periods or injected with periods is irrelevant - the constructor will be whatever it needs to be. How they are _used_ is what is important.

The only problem I have with this current abstract class is that if a class extends `AbstractBooking` and a `PeriodInterface`, it will have to implement `_getPeriod()`, such that it returns `this`. But this also means that it will have to implement `getStart()`, `getEnd()` and `getDuration()`. These three methods are available publically, making the protected versions in this class ... useless.

So the idea of proxying period methods doesn't really make sense. Thus splitting into traits also doesn't make sense. Perhaps an abstract booking implementation doesn't even make sense to begin with. Thoughts?

Comment (XedinUnknown) [reply src/AbstractBooking.php]: My thoughts are that you are over-abstracting. Treat the booking as if it is a period. If there isn't any abstract functionality, then remove the abstract class. Which probably is the case. If agree, then you can still have re-usable functionality in traits, i.e. getters/setters. Certainly, there should be a way to store the status.

Comment (mecha) [reply src/AbstractBooking.php]: I agree.

Note that removing `AbstractBooking` leaves only `StatusAwareTrait` in a repository called `booking-abstract`.

Comment (XedinUnknown) [reply src/AbstractBooking.php]: Well, I gave it as an example of what can be helpful. No reason you can't have a `StartAwareTrait` and `EndAwareTrait` as well. Do you think this is appropriate? How about a `BookingAwareTrait`?

Comment (mecha) [reply src/AbstractBooking.php]: `BookingAwareTrait` is a good idea. I think `StartAwareTrait` and `EndAwareTrait` are more related to periods than bookings; things that what belong in [`rebelcode/time-abstract`](https://github.com/RebelCode/time-abstract).

Comment (XedinUnknown) [reply src/AbstractBooking.php]: Agreed with both points.
